xarray/backends/core.py

```diff
@@ -0,0 +1,425 @@
+
+    def get_name(self):
+        """Return a user-friendly name for prefixing names of dask arrays.
+
+        Not required to be unique.
+        """
+        return ''
+
+
+class OnDiskMixin(object):
+
+    def get_token(self):
+        return (self.filename, os.path.getmtime(self.filename))
+
+    def get_name(self):
+        return self.filename
+
+
+class AbstractWritableDataStore(AbstractDataStore):
+    """An abstract interface for implementing writable datastores."""
+
+    def create_variable(self, name, variable, check_encoding=False):
+        # type: (Hashable, xarray.Variable) -> WritableDuckArray
+        """Create a new variable for writing into the DataStore.
+
+        This method is responsible for setting up a variable to write. It
+        SHOULD NOT actually write array values, but rather create an array to
+        which xarray itself will write.
+
+        If the given variable cannot be stored on the DataStore, this method
+        MUST raise an error.
+
+        Parameters
+        ----------
+        name : Hashable
+            Variable name. If a variable with this name already exists in the
+            DataStore, this method MAY raise an error.
+        variable : xarray.Variable
+            Variable to copy into the DataStore. `variable.encodings` provides
+            a dictionary of DataStore specific options for how to save
+            variables.
+        check_encoding : bool, optional
+            If True, this method SHOULD raise an error for any unexpected keys
+            or invalid values in `variable.encoding`.
+
+        Returns
+        -------
+        Array-like object that writes data to the store when assigning a NumPy
+        array to a tuple of slice objects, e.g., ``x[key] = value``, where
+        ``key`` has type ``Tuple[slice, ...]`` and length equal to the
+        dimensionality of the array, and ``value`` is a ``numpy.ndarray``.
+        """
+        raise NotImplementedError
+
+    def get_writable_array(self, name):
+        # type: (Hashable,) -> WritableDuckArray
+        """Return a writable array corresponding to an existing variable.
+
+        This method is only needed if you want the DataStore to support partial
+        writes, e.g., appending to existing variables.
+
+        Parameters
+        ----------
+        name : Hashable
+            Variable name. How to handle non-existing names is up to the
+            DataStore class. However, xarray will never call this method unless
+            a variable with the given name has already been verified to exist
+            as a member of the mapping returned by `get_variables()`.
+
+        Returns
+        -------
+        Writable array-like object, see `create_variable` for details.
+        """
+        # Note: this mostly exists for the benefit of future support for partial
+        # reads -- we don't actually make use of this in the current version of
+        # xarray.
+        raise NotImplementedError
```

Comment:
This will be passed as a target into `da.store`, so I think it does need to be serializable.